server/src-lib/Hasura/GraphQL/Transport/WebSocket/Server.hs

```diff
@@ -163,17 +182,15 @@ createServerApp
   -- aka WS.ServerApp
   -> WS.PendingConnection
   -> IO ()
-createServerApp (WSServer logger@(L.Logger writeLog) connMap shutdownVar) wsHandlers
-                pendingConn = do
-
+createServerApp (WSServer logger@(L.Logger writeLog) serverStatus) wsHandlers pendingConn = do
   wsId <- WSId <$> UUID.nextRandom
   writeLog $ WSLog wsId EConnectionRequest
-  serverStatus <- STM.readTVarIO shutdownVar
-  case serverStatus of
-    AcceptingConns -> do
+  status <- STM.readTVarIO serverStatus
+  case status of
+    AcceptingConns connMap -> do
       let reqHead = WS.pendingRequest pendingConn
       onConnRes <- _hOnConn wsHandlers wsId reqHead
-      either (onReject wsId) (onAccept wsId) onConnRes
+      either (onReject wsId) (onAccept connMap wsId) onConnRes
```

Comment:
Oh, hm. I didn’t think about the possibility that one can just pull the `connMap` out and hold onto a reference to it—I think there’s still a race here. 😅 The problem is that `serverStatus` might not be `AcceptingCons` anymore by the time `_hOnConn` returns, so the connection should not be accepted.

Fundamentally, the `STMMap.insert` into the connection map *must be atomic* with respect to a check that the server is accepting connections. Otherwise there could always be a race. I’m not sure how to guarantee that at the type level (at least without linear types), since it is, indeed, always possible to hold onto a “stale” connection map this way.

I apologize for the back-and-forth on this. This change seems useful, so I’d definitely like to get it merged! If you’d like, I can make the (small) change I have in mind myself and push it.

Reply:
I'm happy to make the changes, but if you think you have the solution already and will be faster at it, please go ahead. Otherwise point to to the place you think should be changed and I'll do it myself :)

Reply:
Sure, if you’re willing to make the change, that’d be great! I don’t have any super specific solution in mind, but the relevant spot to change is here:

https://github.com/hasura/graphql-engine/blob/768baf6a769b91ccf0e2f5d651569f1ec247799c/server/src-lib/Hasura/GraphQL/Transport/WebSocket/Server.hs#L215

Currently, that uses the `connMap` value that was retrieved earlier, but it needs to instead get the map out of `serverStatus` again to ensure the server hasn’t shut down since then. If it has, we’re in a bit of a tight spot, since we’ve already accepted the connection, so all we can do is immediately close it and bail. Does that make sense?

Reply:
Makes sense!